<a href="https://colab.research.google.com/github/dimasjanotama/data-science-2026/blob/main/Pertemuan10_%5BDimas_Janotama%5D_%5B240401010264%5D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Nama : Dimas Prih P J<br>
NIM : 240401010264<br>
Kelas : IF403<br>

#1. Muat dan Eksplorasi Data
- Pada case ini menggunakan dataset Telco Churn dari Kaggle

In [14]:
import pandas as pd

df = pd.read_csv("telco_churn.csv")
print(df.shape)
print(df["Churn"].value_counts(normalize=True))

(7043, 21)
Churn
No     0.73463
Yes    0.26537
Name: proportion, dtype: float64


#2. Preprocessing
- Encoding pada fitur kategorikal menggunakan pd.get_dummies() atau LabelEncoder
- Bagi data menjadi latih/uji secara stratified untuk menjaga proporsi kelas

In [15]:
from sklearn.model_selection import train_test_split

# TODO: encoding fitur kategorikal (mis. pd.get_dummies)
# TODO: pisahkan X (fitur) dan y (target = Churn)
X_raw = df.drop(columns=["Churn"])
y = df["Churn"]
X = pd.get_dummies(X_raw, drop_first=True)

X_tr, X_te, y_tr, y_te = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42)

#3. Latih Model
- Bangun RandomForestClassifier dengan class_weight="balanced" untuk menangani imbalance

In [16]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(
    n_estimators=300, class_weight="balanced", random_state=42)
rf.fit(X_tr, y_tr)

RandomForestClassifier(class_weight='balanced', n_estimators=300,
                       random_state=42)

#4. Evaluasi
- Hitung precision, recall, F1-score, dan ROC-AUC dengan fokus pada kelas churn (kelas 1)

In [17]:
from sklearn.metrics import classification_report, roc_auc_score

# TODO: hitung prediksi dan probabilitas
# Mengambil prediksi kelas (0 atau 1) untuk classification report
y_pred = rf.predict(X_te)
# Mengambil probabilitas kelas positif (Churn).
y_prob = rf.predict_proba(X_te)[:, 1]

# TODO: tampilkan classification_report dan ROC-AUC
print("================ CLASSIFICATION REPORT ================")
print(classification_report(y_te, y_pred))
print("=======================================================")

roc_auc = roc_auc_score(y_te, y_prob)
print(f"ROC-AUC Score: {roc_auc:.4f}")

================ CLASSIFICATION REPORT ================
              precision    recall  f1-score   support

          No       0.84      0.90      0.87      1035
         Yes       0.65      0.52      0.58       374

    accuracy                           0.80      1409
   macro avg       0.74      0.71      0.72      1409
weighted avg       0.79      0.80      0.79      1409

ROC-AUC Score: 0.8337


#5. Prediksi Probabilitas dan Simpulkan
- Menggunakan predict_proba untuk menghasilkan probabilitas churn tiap pelanggan
- Kesimpulan mengenai temuan utama dari model

In [18]:
# TODO: hitung probabilitas churn (predict_proba)
# Probabilitas untuk kelas positif (Churn)
y_prob_churn = rf.predict_proba(X_te)[:, 1]
print("Array probabilitas churn tiap pelanggan:")
print(y_prob_churn)


# TODO: tulis kesimpulan singkat (3-4 kalimat)

Array probabilitas churn tiap pelanggan:
[0.00333333 0.63666667 0.07333333 ... 0.05333333 0.03       0.01666667]


#Kesimpulan Temuan Utama dari Model:
Model Random Forest yang dibangun berhasil memprediksi perilaku churn pelanggan dengan performa yang baik, ditunjukkan oleh skor ROC-AUC sebesar 0.8337. Secara keseluruhan, model ini sudah siap digunakan untuk mengidentifikasi pelanggan yang berisiko tinggi dan mulai merumuskan strategi retensi.


#Yang Saya Pelajari
Pada pertemuan ini, saya mempelajari alur kerja machine learning untuk prediksi churn menggunakan Pandas dan Scikit-Learn, mulai dari pemisahan data latih dan pengujian menggunakan train_test_split. Selain itu, saya juga mempelajari penanganan data ynag tidak seimbang (imbalanced data) menggunakan parameter class_weight="balanced" pada algoritma Random Forest, serta cara mengevaluasi model menggunakan metrik Classification Report dan skor ROC-AUC.